compiler/rustc_trait_selection/src/traits/select/confirmation.rs

```diff
@@ -834,7 +835,13 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                 obligations.extend(nested);
                 obligations
             })
-            .map_err(|e| OutputTypeParameterMismatch(expected_trait_ref, obligation_trait_ref, e))
+            .map_err(|terr| {
+                OutputTypeParameterMismatch(Box::new(SelectionOutputTypeParameterMismatch {
+                    expected_trait_ref: obligation_trait_ref,
+                    found_trait_ref: expected_trait_ref,
```

Comment:
This looks pretty wrong, except it's not.

Reply:
`expected_trait_ref` in this fn is a misnomer. it comes from taking the self type from an obligation like `fn(): Fn(i32)` and turning it into the `Fn` trait that we know it *does* implement -- in this case, `fn()` implements `Fn()` from its signature. Then we equate `fn(): Fn()` ("expected") and `fn(): Fn(i32)` ("obligation"). But when we're reporting an error, the obligation trait ref is what we want to say was expected from the user's perspective.

we should probably rename the `expected_trait_ref` parameter in this fn.

Reply:
I renamed it to `self_ty_trait_ref`, not sure whether that's the best name.